### PR TITLE
Transient CBS

### PR DESCRIPTION
**Pull request by NoyGabay**
add transient option for CBS,
main difference - use transient A*. 
CBS choosing the best node to expand - this happens naturally when giving CBS SumServiceTimes as a cost function

### CODE REVIEW

File: src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java [hunk 1]
```diff
@@ -178,7 +188,14 @@ private void initOpen(I_ConstraintSet initialConstraints) {
...
     private CBS_Node generateRoot(I_ConstraintSet initialConstraints) {
-        Solution solution = new Solution(); // init an empty solution
+        Solution solution; // init an empty solution
+        if (transientMAPFBehaviour.isTransientMAPF()) {
+            solution = new TransientMAPFSolution();
+        }
+        else {
+            solution = new Solution();
+        }
+
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java:198]:
```suggestion
        // init an empty solution
        Solution solution = transientMAPFBehaviour.isTransientMAPF() ? new TransientMAPFSolution() : new Solution(); 
```
just a little cleaner

File: src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java [hunk 2]
```diff
@@ -281,7 +298,15 @@ private CBS_Node generateNode(CBS_Node parent, Constraint constraint, boolean co
...
-            solution = new Solution(solution);
+//            solution = new Solution(solution);
+
+            if (transientMAPFBehaviour.isTransientMAPF()) {
+                solution = new TransientMAPFSolution(solution);
+            }
+            else {
+                solution = new Solution(solution);
+            }
+
```
Comment (J-morag) [inline src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java:309]:
```suggestion
            solution =  transientMAPFBehaviour.isTransientMAPF() ? new TransientMAPFSolution(solution) : new Solution(solution);
```
a little cleaner

File: src/main/java/Environment/RunManagers/GenericRunManager.java [hunk 3]
```diff
@@ -49,10 +50,16 @@ void setSolvers() {
...
+//        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
+//                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
+//                null, null, null));
+//        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
+//                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
+//                null, null, TransientMAPFBehaviour.transientMAPF));
+        super.solvers.add(new CBS_Solver(null, null, null, null, null, null, null, null, null));
+        super.solvers.add(new CBS_Solver(null, null, null, null, null, null, null, null, TransientMAPFBehaviour.transientMAPF));
+        super.solvers.add(new CBS_Solver(null, null, null, new SumServiceTimes(), null, null, null, null, TransientMAPFBehaviour.transientMAPFsstWithBlacklist));
+
```
Comment (J-morag) [inline src/main/java/Environment/RunManagers/GenericRunManager.java:62]:
```suggestion
        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
                null, null, null));
        super.solvers.add(new CBS_Solver());
```

File: src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java [hunk 4]
```diff
@@ -416,4 +418,27 @@ void sharedGoals(){
...
+        Solution solvedCBSt = CBSt.solve(testInstance, new RunParametersBuilder().setTimeout(1000L).setInstanceReport(instanceReport).createRP());
+        assertTrue(solvedCBSt.solves(testInstance));
+        assertEquals(4 + 3, solvedCBSt.sumIndividualCosts()); // normal SOC function
+        assertEquals(4 + 2, solvedCBSt.sumServiceTimes()); // TMAPF cost function
+        assertEquals(4, solvedCBSt.makespan()); // makespan (normal)
+        assertEquals(4, solvedCBSt.makespanServiceTime()); // makespan (TMAPF)
+
+        System.out.println(solvedNormal);
+        System.out.println(solvedCBSt);
+    }
```
Comment (J-morag) [inline src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java:443]: Can you also add the example from the paper?

Comment (J-morag) [reply src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java:443]: Use a `MapFactory.newArbitraryGraphMap` for it because it isn't on a grid

**Review by J-morag — APPROVED** [review]
A few small comments. good job :)

File: src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java (hunk 4 above)
Comment (NoyGabay) [reply src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java:444]: Done :)

**Comment by NoyGabay** [comment]
Added the example from the paper - looks good. 
I'm wondering if I implemented transientMAPFWithBlacklist correctly in CBS, since I didn't distinguished between it and transientMAPF in the code. transientMAPFsstWithBlacklist acts like them, excluding the sst optimization heuristic.

**Review by J-morag — APPROVED** [review]
For 'transientMAPFWithBlacklist', It should blacklist the targets of all other agents, whenever planning one agent. You can add that, test it, and merge. It would make CBS incomplete, and might not be the best thing to do in CBS, but we need to run experiments and see.
For 'transientMAPFsstWithBlacklist' - basically we have a duplication of indicating sst here and in the cost function parameter. We need to do some refactoring. At the very least we need to make the 'enum' a 'record' to separate out the parameters and make any combination possible. We should probably also forgo having SST indicated there, and instead only use the costFunction interface. It's a bit of a mess. But finish up this PR, and then we can refactor it to be cleaner in another PR.

**Pull request merged**